docs/studies/hidden_linear_function.ipynb

```diff
@@ -0,0 +1,540 @@
+    "\n",
+    "        def vector_in_L(x):\n",
+    "            for y in all_vectors:\n",
+    "                if self.q( (x + y)%2 ) != (self.q(x) + self.q(y))%4:\n",
+    "                    return False\n",
+    "            return True\n",
+    "\n",
+    "        # L is subspace to which we restrict domain of quadratic form.\n",
+    "        # Corresponds to `L_q` in the problem definition.\n",
+    "        self.L = [x for x in all_vectors if vector_in_L(x)]\n",
+    "        \n",
+    "        # All vectors `z` which are solutions to the problem.\n",
+    "        self.all_zs = [z for z in all_vectors if self.is_z(z)]\n",
+    "    \n",
+    "    def is_z(self, z):\n",
+    "        \"\"\"Checks by definition, whether given vector `z` is solution to this problem.\"\"\"\n",
+    "        assert z.shape == (self.n, )\n",
+    "        assert self.L is not None\n",
+    "        for x in self.L:\n",
+    "            if self.q(x) != 2 * ((z @ x) % 2):\n",
+    "                return False\n",
+    "        return True"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For testing, we need to generate an instance of a problem. We can generate random $A$ and $b$. However, for some $A$ and $b$ problem is trivial - that is, $\\mathcal{L}_q = \\{0\\}$ and therefore any $z$ is a solution. In fact, product of $|\\mathcal{L}_q|$ and number of solutions is always equal to $2^n$ (see Lemma 2 in [1]), so we want a problem with large $\\mathcal{L}_q$.\n",
+    "\n",
+    "Code below can be used to generate random problem with given size of $\\mathcal{L}_q$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Size of subspace L: 4\n",
+      "Number of solutions: 256\n"
+     ]
+    }
+   ],
+   "source": [
+    "def random_problem(n, seed=None):\n",
+    "    \"\"\"Generates instance of the problem with given `n`.\n",
+    "    \n",
+    "    Args:\n",
+    "        n: dimension of the problem.\n",
+    "    \"\"\"\n",
+    "    if seed is not None:\n",
+    "        np.random.seed(seed) \n",
+    "    A = np.random.randint(0, 2, size=(n,n))\n",
+    "    for i in range(n):\n",
+    "        for j in range(i+1):\n",
+    "            A[i][j] = 0\n",
+    "    b = np.random.randint(0, 2, size=n)\n",
+    "    problem = HiddenLinearFunctionProblem(A, b)\n",
+    "    return problem\n",
+    "        \n",
+    "def find_interesting_problem(n, min_L_size):\n",
+    "    \"\"\"Generates \"interesting\" instance of the problem.\n",
+    "    \n",
+    "    Args:\n",
+    "        n: dimension of the problem.\n",
+    "        min_L_size: minimal cardinality of subspace L.\n",
+    "        \n",
+    "    Returns instance of problem with gien `n`, such that size of \n",
+    "    subspace `L_q` is at least `min_L_size`.\"\"\"\n",
```

Comment:
I think this description should go before the "Args" section:

```
""Generates "interesting" instance of the problem.

    Returns instance of problem with gien `n`, such that size of 
    subspace `L_q` is at least `min_L_size`.

    Args:
        n: dimension of the problem.
        min_L_size: minimal cardinality of subspace L.
    """
```

Reply:
Done.